gadgets/src/fixed_deposit_tree/builder.rs

```diff
@@ -21,32 +21,49 @@ use sp_std::collections::btree_map::BTreeMap;
 
 #[derive(Clone)]
 pub struct FixedDepositTree {
-	pub depth: usize,
 	secrets: BTreeMap<ScalarBytes, (Scalar, Scalar, Scalar)>,
 	pub hash_params: Poseidon,
 	pub tree: VanillaSparseMerkleTree,
 }
 
 impl FixedDepositTree {
-	pub fn add_secrets(&mut self) -> Scalar {
+	pub fn generate_secrets(&mut self) -> Scalar {
 		let mut rng = OsRng::default();
 		let r = Scalar::random(&mut rng);
 		let nullifier = Scalar::random(&mut rng);
 		let leaf = Poseidon_hash_2(r, nullifier, &self.hash_params);
 		let nullifier_hash = Poseidon_hash_2(nullifier, nullifier, &self.hash_params);
-		self.secrets.insert(leaf.to_bytes(), (r, nullifier, nullifier_hash));
+		self.add_secrets(leaf, r, nullifier, nullifier_hash);
 		leaf
 	}
 
+	pub fn add_secrets(&mut self, leaf: Scalar, r: Scalar, nullifier: Scalar, nullifier_hash: Scalar) {
+		self.secrets.insert(leaf.to_bytes(), (r, nullifier, nullifier_hash));
+	}
+
+	pub fn leaf_data_from_bytes(
+		&self,
+		r_bytes: [u8; 32],
+		nullifier_bytes: [u8; 32],
+	) -> (Scalar, Scalar, Scalar, Scalar) {
+		let r = Scalar::from_bytes_mod_order(r_bytes);
+		let nullifier = Scalar::from_bytes_mod_order(nullifier_bytes);
+		// Construct nullifier hash for note
+		let nullifier_hash = Poseidon_hash_2(nullifier, nullifier, &self.hash_params);
+		// Constructing a leaf from the scalars
+		let leaf = Poseidon_hash_2(r, nullifier, &self.hash_params);
+		(r, nullifier, nullifier_hash, leaf)
+	}
+
 	pub fn get_secrets(&self, leaf: Scalar) -> (Scalar, Scalar, Scalar) {
 		let (r, nullifier, nullifier_hash) = self.secrets.get(&leaf.to_bytes()).unwrap();
 		(*r, *nullifier, *nullifier_hash)
 	}
 
 	pub fn prove_zk(
 		&self,
-		k: Scalar,
 		root: Scalar,
+		leaf: Scalar,
 		bp_gens: &BulletproofGens,
 		mut prover: Prover,
```

Comment:
also here, is this change in the order of the parameters?

Reply:
no I think its change in arguments entirely. Before we specified the index of the leaf, now we use the leaf to get the index.

Reply:
Yes, it does require more memory to store all of these indices. I'm yet to figure out a future proof solution to this

Reply:
We can just save indices of the saved leaves (gotten from the notes) as we insert the leaves, but I need to think about how future clients are going to work